### PR TITLE
added docker

### PR DESCRIPTION
**Pull request by ARK3r**
I added some sections to the README.md as well to explain the usage.

I'm not sure why on my system running with `--net host` it wouldn't work and without this argument, it would work. I'll try to test it on another system later on if no one gets to try it or if anyone knows what the definite way should be.

Also note that in the dockerfile it is cloning my repo. I'm not sure at what point we'd want to swap it so it pulls the `ros2_control_demos` repo.

Also I have humble installed on my system so I had to test it with humble.

Furthermore, I wonder if adding an nvidia version of the script would be useful as well; [something similar to this](https://github.com/ARK3r/f1tenth_hardware_interface/blob/master/run.sh) where at the end it closes the x server as well.

### CODE REVIEW

File: Dockerfile/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  apt-get update && apt-get install -q -y --no-install-recommends \
+    python3-colcon-ros \
+    python3-colcon-common-extensions \
+    python3-rosdep \
+    python3-vcstool \
+    && apt-get clean
+
+RUN mkdir -p /home/ros2_ws/src \
+    && cd /home/ros2_ws/src \
+    && git clone https://github.com/ARK3r/ros2_control_demos -b humble \
```
Comment (moriarty) [inline Dockerfile/Dockerfile:32]: Should this really be doing a git clone here? This is the project getting cloned so a COPY could work and ensure the latest changes locally are picked up
Should this be cloning a fork?
The branch should be gotten from an arg

**Comment by ARK3r** [comment]
@moriarty I have tried to copy the [gz_ros2_control dockerfile](https://github.com/ros-controls/gz_ros2_control/blob/master/Dockerfile/Dockerfile) and make as few changes as possible. I have just about no answer to any of these concerns you've raised except that the gz_ros2_control dockerfile was set that way.

I can make these changes here, and then go and change these on gz_ros2_control as well.

I'll wait for your go signal.

File: Dockerfile/Dockerfile (hunk 1 above)
Comment (ARK3r) [reply Dockerfile/Dockerfile:32]: I think your suggested change makes sense. I'll fix this now.

File: Dockerfile/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
+FROM ubuntu:22.04
```
[the remark this answers is not in the document]
Comment (ARK3r) [reply Dockerfile/Dockerfile:1]: I'm not fully sure why this is the case in [the gz_ros2_control Dockerfile](https://github.com/ros-controls/gz_ros2_control/blob/master/Dockerfile/Dockerfile#L1).

maybe @bmagyar could provide more info as [in this commit](https://github.com/ros-controls/gz_ros2_control/commit/70f3565fee3b9f8e5374baa1380dc8808852cbbb#diff-db7306bd7690f3d533f673528705a4d2be868f4a8c5facd01ae48f8d4a48cc69R1) the change is made and he was involved in the Dockerfiles for that package.

File: Dockerfile/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    dirmngr \
+    gnupg \
+    lsb-release \
+    wget \
+    curl \
+    git \
+    ca-certificates \
+    build-essential \
+    cmake \
+  && apt-get clean
```
[the remark this answers is not in the document]
Comment (ARK3r) [reply Dockerfile/Dockerfile:19]: So do I just remove the `apt-get clean`?

File: Dockerfile/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ARK3r) [reply Dockerfile/Dockerfile:1]: Should I just change this to use the official ros image?

File: Dockerfile/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
```
Comment (moriarty) [inline Dockerfile/Dockerfile:19]:
```suggestion
    cmake
```

File: Dockerfile/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (moriarty) [reply Dockerfile/Dockerfile:19]: Sorry I was travelling yesterday, yes you can just remove the apt-clean and the `\` on the line above

File: Dockerfile/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (moriarty) [reply Dockerfile/Dockerfile:1]: Sorry this wasn't a blocking comment, just a question.

The official image basically does what the first two `RUN apt-get` chunks in this plus install the base ros backage.

**Comment by moriarty** [comment]
> @moriarty I have tried to copy the [gz_ros2_control dockerfile](https://github.com/ros-controls/gz_ros2_control/blob/master/Dockerfile/Dockerfile) and make as few changes as possible. I have just about no answer to any of these concerns you've raised except that the gz_ros2_control dockerfile was set that way.
> 
> I can make these changes here, and then go and change these on gz_ros2_control as well.
> 
> I'll wait for your go signal.

:+1:  sounds good I didn't realise there was another Dockerfile that was being used as a template

File: example_1/bringup/launch/rrbot.launch.py [hunk 5]
```diff
@@ -14,15 +14,24 @@
...
+from launch.substitutions import Command, FindExecutable, LaunchConfiguration, PathJoinSubstitution
 
 from launch_ros.actions import Node
 from launch_ros.substitutions import FindPackageShare
 
 
 def generate_launch_description():
+    start_rviz = LaunchConfiguration("start_rviz")
+
+    start_rviz_arg = DeclareLaunchArgument(
```
Comment (destogl) [inline example_1/bringup/launch/rrbot.launch.py:29]: We usually use different syntay to declare arguments. And also declaring them at the top of the function. Plase check other files and adjust style approprately.

File: README.md [hunk 6]
```diff
@@ -76,6 +76,31 @@ The important files to check in each example are:
 The concepts in this package are demonstrated on the examples of *RRBot* and *DiffBot*.
 Those two world-known imaginary robots are trivial simulations to demonstrate and test `ros2_control` concepts.
 
+## Running
```
Comment (destogl) [inline README.md:79]: Running of what?

this section is all about docker - so we should state this too

File: Dockerfile/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (destogl) [reply Dockerfile/Dockerfile:1]: I propose to use the base image.

**Review by destogl — CHANGES_REQUESTED** [review]

File: Dockerfile/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ARK3r) [reply Dockerfile/Dockerfile:1]: it seems that [osrf/ros doesn't have any base images anymore](https://hub.docker.com/r/osrf/ros/tags?page=1&name=base), so should I just put desktop for now?

File: README.md (hunk 6 above)
Comment (ARK3r) [reply README.md]: Please let me know if this is better

File: Dockerfile/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply Dockerfile/Dockerfile:1]: Did you try https://hub.docker.com/r/osrf/ros/tags ?

**Comment by ARK3r** [comment]
@destogl made changes regarding all reviews

File: Dockerfile/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,22 @@
+ARG ROS_DISTRO=humble
```
Comment (destogl) [inline Dockerfile/Dockerfile:1]: Can we target rolling per default, please?

File: README.md [hunk 8]
```diff
@@ -76,6 +76,33 @@ The important files to check in each example are:
 The concepts in this package are demonstrated on the examples of *RRBot* and *DiffBot*.
 Those two world-known imaginary robots are trivial simulations to demonstrate and test `ros2_control` concepts.
 
+## Usage
+### Using Docker
+**Modifying or building your own**
```
Comment (destogl) [inline README.md:81]:
```suggestion
**Building your own**
```

**Review by destogl — CHANGES_REQUESTED** [review]
Sorry for pickiness. Just 2 tiny things, please 😃

File: Dockerfile/Dockerfile (hunk 7 above)
Comment (ARK3r) [reply Dockerfile/Dockerfile:1]: I was thinking we can change this to rolling when pushing to the rolling branch. I had humble on my system so I developed for humble considering some humble code is running inside the container and I have to run rviz on my own system.

I could try to run the docker inside a docker container with rolling installed though.

**Comment by destogl** [comment]
Regarding merging, let's merge this into master and then backport to humble. That is our usual workflow. Otherwise is hard to keep things synchronized.

File: Dockerfile/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (ARK3r) [reply Dockerfile/Dockerfile]: I changed it so now  it should be using those images

**Comment by destogl** [comment]
> I'll test this inside a rolling docker image and make a new pr to master.

you can update it here and we can update target branch of the PR - maybe needs cherry-picking and force push, but it should be easy.

**Comment by ARK3r** [comment]
I tested docker in docker and it works for rolling.

we can re-target this to master.

**Review by destogl — APPROVED** [review]
Thank you very, very much for all the follow-ups!!

**Comment by christophfroehlich** [comment]
Sorry for being late on this PR, I'm just asking myself while building the docker:
-  Do we need to install ros2_control packages from source here? 
   - If this docker should serve for the demos only -> let's install the released packages with rosdep. 
   - If this docker serves as a basis for contributing to ros2_control, let's use a comple repos file as suggested here https://github.com/ros-controls/control.ros.org/pull/102
- Should the documentation on the usage be published on control.ros.org, or stay in the README only? With gazebo_ros2_control lately, I moved that to the rst for control.ros.org and I have a feeling this would be the better start point for newcomers reading our docs. 
- Let's add a hint on how to launch it with example1
```bash
docker run -it --rm --name ros2_control_demos --net host ros2_control_demos ros2 launch ros2_control_demo_example_1 rrbot.launch.py start_rviz:=false
```

Now having the docker built: I have to add `--net host` to the docker, otherwise I get no messages on the host. @ARK3r does this work for you?

**Comment by ARK3r** [comment]
I think ros2_control_demos should be very low cost for new commers to ros2 control to try. As I just got involved in ros2 control about 2 months ago, it felt very unwelcoming, and I had to continuously push on multiple tutorials to finally be able to set up ros2 control myself. Hopefully having a dockerfile that with one command and little time spent you get a demo going, should help onboard new people.

So on that note, I think we should use rosdep to get the packages so it's fast and I think we should definitely add the docs to control.ros.org.

Regarding the net host arg, I'm not sure what the issue is with my middleware but if I set that arg my memory just keeps getting used up till my PC hangs (oh what a nostalgic event 😂)

**Comment by christophfroehlich** [comment]
@mergifyio backport humble

**Pull request merged**